dockstore-webservice/src/main/java/io/dockstore/webservice/core/LimitedSourceFileBuilder.java

```diff
@@ -0,0 +1,121 @@
+    private static final Logger LOG = LoggerFactory.getLogger(LimitedSourceFileBuilder.class);
+
+    private DescriptorLanguage.FileType type;
+    private String content;
+    private String path;
+    private String absolutePath;
+
+    public LimitedSourceFileBuilder() {
+    }
+
+    public FirstStep start() {
+        return new FirstStep();
+    }
+
+    public class FirstStep extends TypeStep {
+    }
+
+    public class TypeStep {
+        public ContentStep type(DescriptorLanguage.FileType newType) {
+            type = newType;
+            return new ContentStep();
+        }
+    }
+
+    public class ContentStep {
+        public PathStep content(String newContent) {
+            content = newContent;
+            return new PathStep();
+        }
+    }
+
+    public class PathStep {
+        public AbsolutePathStep path(String newPath) {
+            path = newPath;
+            return new AbsolutePathStep();
+        }
+        public BuildStep paths(String newPath) {
+            return path(newPath).absolutePath(newPath);
+        }
+    }
+
+    public class AbsolutePathStep {
+        public BuildStep absolutePath(String newAbsolutePath) {
+            absolutePath = newAbsolutePath;
+            return new BuildStep();
+        }
+    }
+
+    public class BuildStep {
+        public SourceFile build() {
+            SourceFile file = new SourceFile();
+            file.setType(type);
+            setContentWithLimits(file, content, absolutePath);
+            file.setPath(path);
+            file.setAbsolutePath(absolutePath);
+            return file;
+        }
+
+        private static void setContentWithLimits(SourceFile file, String content, String path) {
+            String limitedContent = content;
+            if (content != null) {
+                byte[] bytes = content.getBytes(StandardCharsets.UTF_8);
+                long maximumSize = computeMaximumSize(path);
+                // A large file is probably up to no good.
+                if (bytes.length > maximumSize) {
+                    limitedContent = String.format("Dockstore does not store files of this type over %.1fMB in size", maximumSize / (double) BYTES_PER_MEGABYTE);
+                }
+                // Postgresql cannot store strings that contain "NUL" characters.
+                // https://www.postgresql.org/docs/current/datatype-character.html#DATATYPE-CHARACTER
+                // https://www.ascii-code.com/character/%E2%90%80
+                if (Bytes.indexOf(bytes, Byte.decode("0x00")) != -1) {
+                    limitedContent = "Dockstore does not store binary files";
+                }
+            }
+            file.setContent(limitedContent);
+        }
+
+        private static long computeMaximumSize(String path) {
+            // Jupyter notebooks can contain embedded images, making them larger, on average.
+            if (StringUtils.endsWith(path, ".ipynb")) {
```

Comment:
Think this may still have the problem that the content override is not applied on a notebook entry basis, but on something that looks like a python notebook extension. Could lead to false negatives (if we someday support [rmarkdown](https://bookdown.org/yihui/rmarkdown/notebook.html#notebook-format) notebooks as an example) or false positives (if a malicious workflow imported a bunch of secondary descriptors with the wrong file extension).

Reply:
This feels like a good enough heuristic, because whether it's N `ipynb` or 3N `wdl`, it's still easy to DOS us.  But I can make it more sophisticated if that's really necessary.

Reply:
We can try to detect fakery and other similar stuff, but it doesn't really matter (see Slack discussion).  Using the file suffix is a simple heuristic that will work almost all of the time, and maybe a workflow/notebook will import a big notebook for some valid reason...